### DISCUSSION TITLE
git pull, push seems slow.  Seems to be related to GCM get & store

### DISCUSSION THREAD

**Discussion by rossobianero** [git-ecosystem/git-credential-manager#1258] [General]
I hope I'm posting this in the right place.  I've tried clearing my credentials and reinstalling git.  Has not helped.  I am pulling my hair out here.

Any push or pull (or fetch, etc.) takes at least 40 seconds.  The worst part seems to be two run_command: git-credential-manager calls which alone total over 20 seconds.

I am interacting w/ a Azure DevOps repo using my corporate AD credentials.  I'm on windows 11.  Same behavior in multiple repos using multiple clients (command line, Visual Studio, tortoise git).

Please see the attached trace [gcm-diagnose2.txt](https://github.com/git-ecosystem/git-credential-manager/files/11479404/gcm-diagnose2.txt).  

To highlight the issue, ten seconds elapse between the first two lines:
```
11:15:22.217282 run-command.c:655       trace: run_command: git-credential-manager get
11:15:32.663123 ...\Application.cs:95   trace: [RunInternalAsync] Version: 2.0.935.18315
11:15:32.663123 ...\Application.cs:96   trace: [RunInternalAsync] Runtime: .NET Framework 4.0.30319.42000
11:15:32.663123 ...\Application.cs:97   trace: [RunInternalAsync] Platform: Windows (x86-64)
11:15:32.664124 ...\Application.cs:98   trace: [RunInternalAsync] OSVersion: 10.0 (build 22621)
11:15:32.664124 ...\Application.cs:99   trace: [RunInternalAsync] AppPath: C:\Program Files\Git\mingw64\bin\git-credential-manager
11:15:32.664124 ...\Application.cs:100  trace: [RunInternalAsync] InstallDir: C:\Program Files\Git\mingw64\bin\
```

And about the same amount of time here:
```
11:15:41.931440 run-command.c:655       trace: run_command: git-credential-manager store
11:15:52.217357 ...\Application.cs:95   trace: [RunInternalAsync] Version: 2.0.935.18315
11:15:52.218357 ...\Application.cs:96   trace: [RunInternalAsync] Runtime: .NET Framework 4.0.30319.42000
11:15:52.218357 ...\Application.cs:97   trace: [RunInternalAsync] Platform: Windows (x86-64)
11:15:52.218357 ...\Application.cs:98   trace: [RunInternalAsync] OSVersion: 10.0 (build 22621)
11:15:52.218357 ...\Application.cs:99   trace: [RunInternalAsync] AppPath: C:\Program Files\Git\mingw64\bin\git-credential-manager
11:15:52.218357 ...\Application.cs:100  trace: [RunInternalAsync] InstallDir: C:\Program Files\Git\mingw64\bin\
11:15:52.218357 ...\Application.cs:101  trace: [RunInternalAsync] Arguments: store
```

I found this similar discussion, but I didn't understand the answer and it also doesn't seem 100% related to my problem: https://github.com/git-ecosystem/git-credential-manager/discussions/1169

**Comment by mjcheetham** [comment]
@exbuzz, thanks for the detailed logs!

Looking at the trace, you are correct.. the ~10s startup time is strange.

> I am interacting w/ a Azure DevOps repo using my corporate AD credentials.

I assume you are running on a corporate owned/managed device? Do you know if there are any firewall or anti-virus rules or programs running that may be interfering with GCM starting up? Possibly some sort of scan on each process launch?

**Comment by rossobianero** [comment]
Ok.  I think I found the problem.  It could be a real issue.

I am working on a project must meet [FIPS 140-2 validation requirements](https://learn.microsoft.com/en-us/windows/security/threat-protection/fips-140-validation).  We have had "[FIPS mode](https://docs.logrhythm.com/docs/enterprise/client-console-administrator-guide/deployment-security/federal-information-processing-standards-fips/put-windows-into-fips-mode)"  enabled on our lab sever for integration testing.  At some point I enabled it on my dev box, just to make sure we caught any issues like that as early as possible.  "What could be the harm...?"

As part of trying to get to the bottom of this issue, I tried running "C:\Program Files\Git\mingw64\bin\git-credential-manager" directly (with no arguments).  The experience was:

1) Run the command
2) Nothing happens for 8 seconds
3) I then see that first trace line I mentioned above:  
 ...\Application.cs:95   trace: [RunInternalAsync] Version: 2.0.935.18315
Followed by the the help screen from git-credential-manager.exe and the warning "Required command was not provided."

This all seem expected to me.  And that 8 second startup time seemed like a possible contributing factor to my issues.

Then, I disabled FIPS mode.  

That command now runs instantly without delay.  So I went back and did my pull...now 1.13 seconds down from ~40 seconds....
